File: day1/1.c

#include "../a.h"

SI p1(C*z) {
    I f=0,l=strlen(z);
    W(!isdigit(z[f])&&z[f])f++;
    W(!isdigit(z[l])&&l)l--;
    R 10*(z[f]-'0')+(z[l]-'0'); }

SI st(C*w,C*z) {W(*w)IF(*w!=*z,R 0)ELSE(w++,z++)R 1;}

SI num(C*z) {
    if(st("one",z))R 1;
    if(st("two",z))R 2;
    if(st("three",z))R 3;
    if(st("four",z))R 4;
    if(st("five",z))R 5;
    if(st("six",z))R 6;
    if(st("seven",z))R 7;
    if(st("eight",z))R 8;
    if(st("nine",z))R 9;
    if(isdigit(*z))R*z-'0';
    R-1; }

SI p2(C*z) {
    I f=-1,l=-1,d;
    W(*z) {
        d=num(z++);
        IF(d!=-1,IF(f==-1,f=d)ELSE(l=d)) }
    R 10*f+(l==-1?f:l);
}

I main(V) {
    FILE*in=fopen("input.txt","r");
    I s1=0,s2=0;
    W(!feof(in)) {
        C buf[128]={0};fgets(buf,128,in);
        IF(!*buf,continue)
        s1+=p1(buf),s2+=p2(buf); }
    printf("%d\n%d\n", s1, s2);
}

File: day1/1.hs
import Data.Maybe (mapMaybe, listToMaybe, fromMaybe, fromJust)
import Data.List (isPrefixOf, elemIndex)

readLines :: FilePath -> IO [String]
readLines file = do
  contents <- readFile file
  return $ lines contents

-- Part 1
firstDigit :: String -> Char
firstDigit = head . filter (`elem` ['0'..'9'])

part1 :: [String] -> Int
part1 lines = sum (map (\x -> read [firstDigit x, firstDigit $ reverse x]) lines)

-- Part 2
firstNumber :: String -> [String] -> Int
firstNumber x d = fromMaybe (firstNumber (tail x) d) p
    where
        n = listToMaybe $ mapMaybe (\z -> if z `isPrefixOf` x then Just z else Nothing) d
        p = fmap (\x -> 1 + fromJust (x `elemIndex` d) `mod` 9) n

part2 :: [String] -> Int
part2 = sum . map (\line -> 10 * firstNumber line numbers + firstNumber (reverse line) (map reverse numbers))
    where numbers = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"] ++ [show x | x <- [1 .. 9]]

main = do
  lines <- readLines "input.txt"
  print $ part1 lines
  print $ part2 lines